Rework Random

flaviojs opened the pull request:
1. Use C++ facilities to generate uniform pseudo-random integer numbers.
2. Replace leftover uses of `rand()` with `Random`.
3. Add game policy to allow not recycling prerandom numbers.
    Vanilla had this commented out in the name of optimization, but today computers are much more powerful and this is not a hot path so I don't see value in this optimization.

This PR is based on the opposite view of #777.
For me non-uniform pseudo-random numbers is a bug (`PreRandom` was not uniform) and issue #777 is not a bug.

lynxlynxlynx commented:
LGTM, but I'd go one step further. Since the original prerandom range is nowadays much larger, I see no point in the extra setting. I'd also remove prerandom and prechance completely, since the speed difference should be negligible.

flaviojs commented:
I wouldn't remove `PreRandom/PreChance`. It provides some predictability with save-scumming during combat, which is a valid strategy.

As for the new extra setting, I'll gladly remove it if we decide to not recycle prerandom numbers.
@selaux your thoughs?

lynxlynxlynx commented:
The setting would then better suit this switch: whether to use prerandom at all or redirect to random. :)

selaux commented:
Had a look at the PR and same as @lynxlynxlynx, I would remove the switch, never recycling numbers sounds good to me. I would not even add a switch between random and prerandom, as I don't really see a good usecase for that (except for saveloading yourself through the game :wink:)

flaviojs commented:
Removed the game policy and reworked the commits for clarity.

@selaux I didn't understand, do you also want direct random values to be used instead of pregenerated values in PreRandom?
I'm not into that but I'll implement it if both of you agree it should be like that.

Btw, when this is merged #777 should lose the `bug` label and be closed with  a `wontfix` label.
It's an "either-or" situation.

lynxlynxlynx commented on src/sgp/Random.cc:
```diff
@@ -1,69 +1,89 @@
 
+/// Pre-generated pseudo-random number engine.
+struct PreRandomEngine {
+	typedef UINT32 result_type;
```
I'd remove this typedef, since it doesn't add much (pretty obvious the types should match) and max() uses a fixed constant anyway.

flaviojs replied:
Can't be removed, `std::uniform_int_distribution` requires that typedef and the other member functions.

selaux commented:
@flaviojs I just wanted to emphasize that there is a reason for having random and prerandom. Other than that I think it is fine now...
